pkgs/development/compilers/openjdk/generic.nix

```
@@ -0,0 +1,660 @@
+    ''
+    + lib.optionalString (!atLeast11) ''
+      ln -s $jre/lib/openjdk/jre/bin $jre/bin
+      ln -s $jre/lib/openjdk/jre $out/jre
+    '';
+
+  preFixup =
+    (
+      if atLeast11 then
+        ''
+          # Propagate the setJavaClassPath setup hook so that any package
+          # that depends on the JDK has $CLASSPATH set up properly.
+          mkdir -p $out/nix-support
+          #TODO or printWords?  cf https://github.com/NixOS/nixpkgs/pull/27427#issuecomment-317293040
+          echo -n "${setJavaClassPath}" > $out/nix-support/propagated-build-inputs
+        ''
+      else
+        ''
+          # Propagate the setJavaClassPath setup hook from the JRE so that
+          # any package that depends on the JRE has $CLASSPATH set up
+          # properly.
+          mkdir -p $jre/nix-support
+          printWords ${setJavaClassPath} > $jre/nix-support/propagated-build-inputs
+        ''
+    )
+    + ''
+
+      # Set JAVA_HOME automatically.
+      mkdir -p $out/nix-support
+      cat <<EOF > $out/nix-support/setup-hook
+      if [ -z "\''${JAVA_HOME-}" ]; then export JAVA_HOME=$out/lib/openjdk; fi
+      EOF
+    '';
+
+  postFixup = ''
+    # Build the set of output library directories to rpath against
+    LIBDIRS=""
+    for output in $(getAllOutputNames); do
+      if [ "$output" = debug ]; then continue; fi
+      LIBDIRS="$(find $(eval echo \$$output) -name \*.so\* -exec dirname {} \+ | ${
+        if atLeast17 then "sort -u" else "sort | uniq"
+      } | tr '\n' ':'):$LIBDIRS"
+    done
+    # Add the local library paths to remove dependencies on the bootstrap
+    for output in $(getAllOutputNames); do
+      if [ "$output" = debug ]; then continue; fi
+      OUTPUTDIR=$(eval echo \$$output)
+      BINLIBS=$(find $OUTPUTDIR/bin/ -type f; find $OUTPUTDIR -name \*.so\*)
+      echo "$BINLIBS" | while read i; do
+        patchelf --set-rpath "$LIBDIRS:$(patchelf --print-rpath "$i")" "$i" || true
+        patchelf --shrink-rpath "$i" || true
+      done
+    done
+  '';
+
+  # TODO: The OpenJDK 8 derivation got this wrong.
+  disallowedReferences = [
+    (if atLeast11 then jdk-bootstrap' else jdk-bootstrap)
+  ];
+
+  passthru =
+    {
+      home = "${finalAttrs.finalPackage}/lib/openjdk";
+      inherit jdk-bootstrap;
+      inherit (source) updateScript;
+    }
+    // (if atLeast11 then { inherit gtk3; } else { inherit gtk2; })
+    // lib.optionalAttrs (!atLeast23) {
+      inherit architecture;
+    };
+
+  meta = {
+    description = "Open-source Java Development Kit";
+    homepage = "https://openjdk.java.net/";
+    license = lib.licenses.gpl2Only;
+    maintainers = with lib.maintainers; [
+      edwtjo
+      infinidoge
+    ];
+    mainProgram = "java";
+    platforms =
+      [
+        "i686-linux"
+        "x86_64-linux"
+        "aarch64-linux"
+      ]
+      ++ lib.optionals atLeast11 [
+        "armv7l-linux"
+        "armv6l-linux"
+        "powerpc64le-linux"
+      ];
```

```suggestion
      ++ lib.optionals atLeast17 [
        "riscv64-linux"
      ];
```

This would make https://github.com/NixOS/nixpkgs/pull/343907 obsolete.

---

Would you be up for rebasing your PR after this one merges? I’d be happy to review and merge it then (though I don’t have hardware to test it out). I’ve tried to keep this PR strictly to changes that result in no change in the resulting derivations and `meta`s here in order to separate *organizing the current state of things* from *deciding what the state of things ought to be*; keeping it purely mechanical means that this PR can’t be responsible for any regressions, should never need to be reverted, and doesn’t express any policy other than “copy‐pasted code is bad”. Not that I expect adding RISC‐V support would lead to any of those, which is why I’m happy to hit the merge button after :)

---

Sure, no problem.